### PR TITLE
Update cordova-plugin-ble-central to 1.2.2 (latest)

### PR DESCRIPTION
**Pull request by peitschie**
Types were found via direct code inspection of plugin, primarily via the Android implementation at https://github.com/don/cordova-plugin-ble-central/blob/ac55995dc9e9d19f231c13e6bbfb9ba24b076480/src/android/BLECentralPlugin.java and https://github.com/don/cordova-plugin-ble-central/blob/ac55995dc9e9d19f231c13e6bbfb9ba24b076480/src/android/Peripheral.java

Changes also picked out of https://github.com/don/cordova-plugin-ble-central/blob/ac55995dc9e9d19f231c13e6bbfb9ba24b076480/CHANGES.txt

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:
If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/don/cordova-plugin-ble-central & https://github.com/don/cordova-plugin-ble-central/blob/master/src/android/BLECentralPlugin.java
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/cordova-plugin-ble-central/index.d.ts [hunk 1]
```diff
@@ -35,74 +41,105 @@ declare namespace BLECentralPlugin {
...
+        requestMtu(device_id: string, mtu: number): void;
+        requestMtu(device_id: string, mtu: number, success: () => any): void;
+        requestMtu(device_id: string, mtu: number, success: () => any, failure: () => any): void;
+
+        /* Clears cached services and characteristics info for some poorly behaved devices. Uses an undocumented API,
+        so it is not guaranteed to work.
+           [iOS] refreshDeviceCache is not supported on iOS. */
+        refreshDeviceCache(device_id: string, timeout_millis: number): void;
+        refreshDeviceCache(device_id: string, timeout_millis: number, success: (data: PeripheralDataExtended) => any): void;
+        refreshDeviceCache(device_id: string, timeout_millis: number, success: (data: PeripheralDataExtended) => any, failure: (error: string|BLEError) => any): void;
```
Comment (DanielRosenwasser) [inline types/cordova-plugin-ble-central/index.d.ts]: Combine these 3 overloads into a single signature with optional parameters.

File: types/cordova-plugin-ble-central/index.d.ts [hunk 2]
```diff
@@ -35,74 +41,105 @@ declare namespace BLECentralPlugin {
...
         /* Calls the success callback when the peripheral is connected and the failure callback when not connected. */
         isConnected(device_id: string, success: () => any): void;
-        isConnected(device_id: string, success: () => any, failure: () => any): void;
+        isConnected(device_id: string, success: () => any, failure: (error: string) => any): void;
+
+        /* May be used to request (on Android) a larger MTU size to be able to send more data at once
+           [iOS] requestMtu is not supported on iOS. */
+        requestMtu(device_id: string, mtu: number): void;
+        requestMtu(device_id: string, mtu: number, success: () => any): void;
+        requestMtu(device_id: string, mtu: number, success: () => any, failure: () => any): void;
```
Comment (DanielRosenwasser) [inline types/cordova-plugin-ble-central/index.d.ts]: Combine these 3 overloads into a single signature with optional parameters.

File: types/cordova-plugin-ble-central/index.d.ts [hunk 3]
```diff
@@ -35,74 +41,105 @@ declare namespace BLECentralPlugin {
...
         /* Register to be notified when the value of a characteristic changes. */
         startNotification(device_id: string, service_uuid:string, characteristic_uuid:string, success: (rawData: ArrayBuffer) => any): void;
-        startNotification(device_id: string, service_uuid:string, characteristic_uuid:string, success: (rawData: ArrayBuffer) => any, failure: () => any): void;
-        
+        startNotification(device_id: string, service_uuid:string, characteristic_uuid:string, success: (rawData: ArrayBuffer) => any, failure: (error: string|BLEError) => any): void;
+
         stopNotification(device_id: string, service_uuid:string, characteristic_uuid:string): void;
         stopNotification(device_id: string, service_uuid:string, characteristic_uuid:string, success: () => any): void;
-        stopNotification(device_id: string, service_uuid:string, characteristic_uuid:string, success: () => any, failure: () => any): void;
+        stopNotification(device_id: string, service_uuid:string, characteristic_uuid:string, success: () => any, failure: (error: string|BLEError) => any): void;
```
Comment (DanielRosenwasser) [inline types/cordova-plugin-ble-central/index.d.ts]: Combine these 3 overloads into a single signature with optional parameters.

File: types/cordova-plugin-ble-central/index.d.ts [hunk 4]
```diff
@@ -35,74 +41,105 @@ declare namespace BLECentralPlugin {
...
+        write(device_id: string, service_uuid:string, characteristic_uuid:string, data: ArrayBuffer, success  : () => any, failure: (error: string|BLEError) => any): void;
 
-        /* Writes data to a characteristic without a response from the peripheral. You are not notified if the write fails in the BLE stack. 
+        /* Writes data to a characteristic without a response from the peripheral. You are not notified if the write fails in the BLE stack.
         The success callback is be called when the characteristic is written.*/
         writeWithoutResponse(device_id: string, service_uuid:string, characteristic_uuid:string, data: ArrayBuffer): void;
         writeWithoutResponse(device_id: string, service_uuid:string, characteristic_uuid:string, data: ArrayBuffer, success : () => any): void;
-        writeWithoutResponse(device_id: string, service_uuid:string, characteristic_uuid:string, data: ArrayBuffer, success : () => any, failure: () => any): void;
+        writeWithoutResponse(device_id: string, service_uuid:string, characteristic_uuid:string, data: ArrayBuffer, success : () => any, failure: (error: string) => any): void;
```
Comment (DanielRosenwasser) [inline types/cordova-plugin-ble-central/index.d.ts]: Combine these 3 overloads into a single signature with optional parameters.

File: types/cordova-plugin-ble-central/index.d.ts [hunk 5]
```diff
@@ -35,74 +41,105 @@ declare namespace BLECentralPlugin {
...
 
         read(device_id: string, service_uuid:string, characteristic_uuid:string): void;
         read(device_id: string, service_uuid:string, characteristic_uuid:string, success: (rawData: ArrayBuffer) => any): void;
-        read(device_id: string, service_uuid:string, characteristic_uuid:string, success: (rawData: ArrayBuffer) => any, failure: () => any): void;
+        read(device_id: string, service_uuid:string, characteristic_uuid:string, success: (rawData: ArrayBuffer) => any, failure: (error: string|BLEError) => any): void;
 
         write(device_id: string, service_uuid:string, characteristic_uuid:string, data: ArrayBuffer): void;
         write(device_id: string, service_uuid:string, characteristic_uuid:string, data: ArrayBuffer, success  : () => any): void;
-        write(device_id: string, service_uuid:string, characteristic_uuid:string, data: ArrayBuffer, success  : () => any, failure: () => any): void;
+        write(device_id: string, service_uuid:string, characteristic_uuid:string, data: ArrayBuffer, success  : () => any, failure: (error: string|BLEError) => any): void;
```
Comment (DanielRosenwasser) [inline types/cordova-plugin-ble-central/index.d.ts]: Combine these 3 overloads into a single signature with optional parameters.

File: types/cordova-plugin-ble-central/index.d.ts [hunk 6]
```diff
@@ -35,74 +41,105 @@ declare namespace BLECentralPlugin {
...
 
         disconnect(device_id:string): void;
         disconnect(device_id:string, success: () => any): void;
-        disconnect(device_id:string, success: () => any, failure: () => any): void;
+        disconnect(device_id:string, success: () => any, failure: (error: string|BLEError) => any): void;
 
         read(device_id: string, service_uuid:string, characteristic_uuid:string): void;
         read(device_id: string, service_uuid:string, characteristic_uuid:string, success: (rawData: ArrayBuffer) => any): void;
-        read(device_id: string, service_uuid:string, characteristic_uuid:string, success: (rawData: ArrayBuffer) => any, failure: () => any): void;
+        read(device_id: string, service_uuid:string, characteristic_uuid:string, success: (rawData: ArrayBuffer) => any, failure: (error: string|BLEError) => any): void;
```
Comment (DanielRosenwasser) [inline types/cordova-plugin-ble-central/index.d.ts]: Combine these 3 overloads into a single signature with optional parameters.

File: types/cordova-plugin-ble-central/index.d.ts [hunk 7]
```diff
@@ -35,74 +41,105 @@ declare namespace BLECentralPlugin {
...
+
+        /* Automatically connect to a device when it is in range of the phone
+           [iOS] background notifications on ios must be enabled if you want to run in the background
+           [Android] this relies on the autoConnect argument of BluetoothDevice.connectGatt(). Not all Android devices implement this feature correctly. */
+        autoConnect(device_id:string, success: (data: PeripheralDataExtended) => any, failure: (error: string|BLEError) => any): void;
 
         disconnect(device_id:string): void;
         disconnect(device_id:string, success: () => any): void;
-        disconnect(device_id:string, success: () => any, failure: () => any): void;
+        disconnect(device_id:string, success: () => any, failure: (error: string|BLEError) => any): void;
```
Comment (DanielRosenwasser) [inline types/cordova-plugin-ble-central/index.d.ts]: Combine these 3 overloads into a single signature with optional parameters.

File: types/cordova-plugin-ble-central/index.d.ts [hunk 8]
```diff
@@ -35,74 +41,105 @@ declare namespace BLECentralPlugin {
...
-        
+        scan(services: string[], seconds: number, success : (data: PeripheralData) => any, failure : (error: string) => any): void;
+
         startScan(services: string[], success: (data: PeripheralData) => any): void;
-        startScan(services: string[], success: (data: PeripheralData) => any, failure: () => any): void;
+        startScan(services: string[], success: (data: PeripheralData) => any, failure: (error: string|BLEError) => any): void;
 
         startScanWithOptions(services: string[], options: StartScanOptions, success: (data: PeripheralData) => any): void;
-        startScanWithOptions(services: string[], options: StartScanOptions, success: (data: PeripheralData) => any, failure: () => any): void;
+        startScanWithOptions(services: string[], options: StartScanOptions, success: (data: PeripheralData) => any, failure: (error: string) => any): void;
```
Comment (DanielRosenwasser) [inline types/cordova-plugin-ble-central/index.d.ts]: Combine these 2 overloads into a single signature with optional parameters.

File: types/cordova-plugin-ble-central/index.d.ts [hunk 9]
```diff
@@ -35,74 +41,105 @@ declare namespace BLECentralPlugin {
 
     export interface BLECentralPluginStatic {
         scan(services: string[], seconds: number, success : (data: PeripheralData) => any): void;
-        scan(services: string[], seconds: number, success : (data: PeripheralData) => any, failure : () => any): void;
-        
+        scan(services: string[], seconds: number, success : (data: PeripheralData) => any, failure : (error: string) => any): void;
```
Comment (DanielRosenwasser) [inline types/cordova-plugin-ble-central/index.d.ts]: Combine these 2 overloads into a single signature with optional parameters.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
Thanks! Can you make the requested changes to the overloads?

**Comment by peitschie** [comment]
Done! Sorry for the bit of a time waste there @DanielRosenwasser ... I probably should have looked at those first rather than following the existing pattern.

Thanks for the review 😺

**Pull request merged**